Add is-triangle exercise + solution

vaibhavarora14 opened the pull request:
This solution is based on `Triangle inequality theorem`, which states
sum of any 2 sides of triangle, should be greater than 3rd one.

![triangle-inequality-theorem](https://user-images.githubusercontent.com/24986539/80281035-fb6d3780-8725-11ea-8064-0c0a15625425.png)

Commit:
Add is-triangle exercise + solution

ArfatSalman commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,53 @@
+  let result = true;
+
+  for (const arg of args) {
+    if (arg !== Math.abs(arg)) {
```
Since the function is `isPositiveInteger`, won't the `args` be numbers?

ArfatSalman replied:
Also, why not this?
```suggestion
    if (arg < 0) {
```

ArfatSalman commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,53 @@
+function isPositiveInteger(...args) {
+  let result = true;
+
+  for (const arg of args) {
```
Would this work?
```suggestion
args.every(arg => arg > 0)
```

varora1406 replied:
Yes, changed in latest commit with `every`

ArfatSalman commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,53 @@
+function isNumber(...args) {
+  let result = true;
+
+  for (const arg of args) {
```
```suggestion
args.every(arg => typeof === 'number')
```

ArfatSalman commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,53 @@
+  for (const arg of args) {
+    if (arg !== Math.abs(arg)) {
+      result = false;
+      break;
```
`break` is not a good idea at all.

varora1406 replied:
As we found alternate, so this is not longer used now

ArfatSalman commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,53 @@
+}
+
+function isPositiveInteger(...args) {
+  let result = true;
```
Not required when you use array functions.

varora1406 replied:
correct, no longer exist in latest code pushed

vaibhavarora14 committed:
Replace Math.abs with check of "number > 0"

Co-Authored-By: Arfat Salman <giney.paradise@gmail.com>

Commit:
Update program with some betterments, which are :
- Using array utilities to reduce LOC in number utilities created in this file
- change some naming

Commit:
Merge branch 'is-triangle' of https://github.com/pesto-students/deep-dive-varora1406 into is-triangle

varora1406 commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,53 @@
+
+  for (const arg of args) {
+    if (typeof arg !== 'number') {
+      result = false;
```
Right, have a look at latest code. This block has changed and sort of removed.

varora1406 commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,53 @@
+  for (const arg of args) {
+    if (arg !== Math.abs(arg)) {
+      result = false;
+      break;
```
yes, it's not in latest code now.
Have a look at it. 🙂

ahhda commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,25 @@
+    throw Error(`program expect all sides positive numbers, instead got ${side1}, ${side2}, ${side3}`);
+  }
+
+  if (side1 + side2 <= side3) {
+    return false;
+  }
+
+  if (side2 + side3 <= side1) {
+    return false;
+  }
+
+  return (side3 + side1 <= side2);
```
This kind of breaks my mental model and makes it look like you are giving some kind of preference to some sides. Again not a required change, but something like this makes it look better

```js
const sumofAB = sideA + sideB;
const sumofBC = sideB + sideC;
const sumofCA = sideC + sideA;
return (sumofAB > sideC && sumofBC > sideA && sumofCA > sideB)
```

varora1406 replied:
done 6e34cd1bb192b07eded55b418f0012cafcbc00ff
Updating subject of PR as well soon

ahhda commented on js-exercises/is-triangle/isTriangle.js:
```diff
@@ -0,0 +1,25 @@
+const isNumber = (arg) => typeof arg === 'number';
+
+const isPositiveNumber = (arg) => isNumber(arg) && (arg > 0);
+
+const arePositiveNumbers = (args) => args.every(isPositiveNumber);
```
Great use of predicates! 👍

ahhda commented on js-exercises/.eslintrc:
```diff
@@ -6,6 +6,7 @@
     "no-restricted-syntax": 0,
     "arrow-parens": 0,
-    "import/prefer-default-export": 0
+    "import/prefer-default-export": 0,
+    "linebreak-style": 0
```
As mentioned in the guidelines before, if you are making changes to ESlint or something which is directly not related to the question, please create a separate PR for that.

varora1406 replied:
right, will take care further

ahhda reviewed (changes requested):
Good work! Just a reminder, separate PR's for different kinds of changes.

Commit:
add variables for operations output for
- better readability of code

Commit:
fix issue of 'args' not being array of params

vaibhavarora14 commented:
> 
> 
> Good work! Just a reminder, separate PR's for different kinds of changes.

Will take care from next time 👍
